cpu/stm32f3/periph/i2c.c

```diff
@@ -0,0 +1,490 @@
+    i2c->TXDR = reg;
+
+    /* write out data bytes */
+    _write(i2c, data, length);
+
+    /* end transmission */
+    _stop(i2c);
+
+    return length;
+}
+
+void i2c_poweron(i2c_t dev)
+{
+    switch (dev) {
+#if I2C_0_EN
+        case I2C_0:
+            I2C_0_CLKEN();
+            break;
+#endif
+    }
+}
+
+void i2c_poweroff(i2c_t dev)
+{
+    switch (dev) {
+#if I2C_0_EN
+        case I2C_0:
+            while (I2C_0_DEV->ISR & I2C_ISR_BUSY);
+
+            I2C_0_CLKDIS();
+            break;
+#endif
+    }
+}
+
+static void _start(I2C_TypeDef *dev, uint8_t address, uint8_t length, uint8_t rw_flag)
+{
+    /* set address mode to 7-bit */
+    dev->CR2 &= ~(I2C_CR2_ADD10);
+
+    /* set slave address */
+    dev->CR2 &= ~(I2C_CR2_SADD);
+    dev->CR2 |= (address << 1);
+
+    /* set transfer direction */
+    dev->CR2 &= ~(I2C_CR2_RD_WRN);
+    dev->CR2 |= (rw_flag << 10);
+
+    /* set number of bytes */
+    dev->CR2 &= ~(I2C_CR2_NBYTES);
+    dev->CR2 |= (length << 16);
+
+    /* configure autoend configuration */
+    dev->CR2 &= ~(I2C_CR2_AUTOEND);
+
+    /* generate start condition */
+    DEBUG("Generate start condition\n");
+    dev->CR2 |= I2C_CR2_START;
+}
+
+static inline void _read(I2C_TypeDef *dev, char *data, int length)
+{
+    for (int i = 0; i < length; i++) {
+        /* wait for transfer to finish */
+        DEBUG("Waiting for DR to be full\n");
+        while (!(dev->ISR & I2C_ISR_RXNE));
+        DEBUG("DR is now full\n");
+
+        /* read data from data register */
+        data[i] = (uint8_t)dev->RXDR;
+        DEBUG("Read byte %i from DR\n", i);
+    }
+}
+
+static inline void _write(I2C_TypeDef *dev, char *data, int length)
+{
+    for (int i = 0; i < length; i++) {
+        /* wait for ack */
+        DEBUG("Waiting for ACK\n");
+        while (!(dev->ISR & I2C_ISR_TXIS));
+
+        /* write data to data register */
+        DEBUG("Write byte %i to DR\n", i);
+        dev->TXDR = (uint8_t)data[i];
+        DEBUG("Sending data\n");
+    }
+}
+
+static inline void _stop(I2C_TypeDef *dev)
+{
+    /* make sure transfer is complete */
+    DEBUG("Wait for transfer to be complete\n");
+    while (!(dev->ISR & I2C_ISR_TC));
+
+    /* send STOP condition */
+    DEBUG("Generate stop condition\n");
+    dev->CR2 |= I2C_CR2_STOP;
+}
+
+#if I2C_0_EN
+void I2C_0_ERR_ISR(void)
+{
+    unsigned state = I2C_0_DEV->ISR;
+    DEBUG("\n\n### I2C ERROR OCCURED ###\n");
+    DEBUG("status: %08x\n", state);
+    if (state & I2C_ISR_OVR) {
+        DEBUG("OVR\n");
+    }
+    if (state & I2C_ISR_NACKF) {
+        DEBUG("AF\n");
+    }
+    if (state & I2C_ISR_ARLO) {
+        DEBUG("ARLO\n");
+    }
+    if (state & I2C_ISR_BERR) {
+        DEBUG("BERR\n");
+    }
+    if (state & I2C_ISR_PECERR) {
+        DEBUG("PECERR\n");
+    }
+    if (state & I2C_ISR_TIMEOUT) {
+        DEBUG("TIMEOUT\n");
+    }
+    if (state & I2C_ISR_ALERT) {
+        DEBUG("SMBALERT\n");
+    }
+    while (1);
```

Comment:
this seems dangerous... If you really want to stop here, better call `core_panic()`, this way the board will at least reset if `DEVELHELP` is not set. Otherwise I think it makes more sense to either leave out the handling of this interrupt all together or really handle the errors (e.g. by resetting the device...).